### PR TITLE
Update CDK from v1 to v2

### PR DESCRIPTION
**Pull request by pergardebrink**
# Why?

CDKv1 is going out of support and should be replaced. For us, we need to be able to use the new STANDARD_6_0 CodeBuild image as part of our pipelines. This cannot be defined in CDKv1, so it has to be upgraded to CDKv2

*Issue #, if available:*
https://github.com/awslabs/aws-deployment-framework/issues/617
https://github.com/awslabs/aws-deployment-framework/issues/503
https://github.com/awslabs/aws-deployment-framework/issues/614

## What?

Description of changes:
- Bumped CDK to version 2.68 (latest CDKv2)
- Updated code according to the CDKv1 to v2 migration guide: https://docs.aws.amazon.com/cdk/v2/guide/migrating-v2.html
- Updated the default CodeBuild image since the old (UBUNTU_14_04_PYTHON_3_7_1) was unsupported and not available in CDKv2
- Use BootstraplessSynthesizer (https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk/BootstraplessSynthesizer.html) to avoid having to add permissions for ssm:GetParameters because of the bootstrap check (this should be ok for now unless the PipelineManagementCodeBuild project would need to deploy any CDK assets)

---

By submitting this pull request, I confirm that you can use, modify, copy, and
redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py
Comment (sbkok) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py:25]: Yes, we should get that updated.
The reason we didn't was not to introduce a breaking change.

Considering CDK v2 would be a breaking change anyway we can fix it here.
I would ideally like to move this to the `adfconfig.yml` configuration file. So we can suggest customers to update it from their config instead. So we don't need to introduce a breaking change later if our default value is no longer supported.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_events.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_events.py:14]:
```suggestion
    Stack,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codepipeline.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codepipeline.py:15]:
```suggestion
    Fn,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_chatbot.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_chatbot.py:11]: Can you add a comma at the end, so in the future if we add more requirements we don't need to modify this line too?

```suggestion
    Stack,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py:17]:
```suggestion
    Aws,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codecommit.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codecommit.py:9]:
```suggestion
    aws_codepipeline as _codepipeline,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codestar.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codestar.py:10]:
```suggestion
    aws_codepipeline as _codepipeline,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_github.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_github.py:11]:
```suggestion
    Token,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_jenkins.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_jenkins.py:9]:
```suggestion
    aws_codepipeline as _codepipeline,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_notifications.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_notifications.py:14]:
```suggestion
    Stack,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_s3.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_s3.py:10]:
```suggestion
    aws_codepipeline as _codepipeline,
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/generate_pipeline_stacks.py
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/generate_pipeline_stacks.py:37]: Thanks for adding this!

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/requirements.txt
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/requirements.txt:3]: You gotta love the new requirements list with CDK v2 haha, I love it! So short :D

**Review by sbkok — CHANGES_REQUESTED** [review]
Thank you for contributing this.
I understand the urgent need for this, looks like we need to go ahead and release v4 earlier than I anticipated.

A few minor changes are required, mostly around trailing commas. This would make it easier to merge and auto-merge PRs that add new imports later.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py
Comment (pergardebrink) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py:25]: Yes, that would make sense! 

Should we move to have this configuration setting into a separate issue/pull request outside the scope of this pull request as we are probably not targeting 3.x and will inevitably be introducing a breaking change (and perhaps try to make it configurable before 4.0)? 

I could potentially look into that, but I have only dug enough into the ADF codebase to be able to do the CDKv1->CDKv2 upgrade, but could probably find my way to work with the adfconfig.yml (maybe be pointed in the right direction).

**Comment by pergardebrink** [comment]
> Thank you for contributing this. I understand the urgent need for this, looks like we need to go ahead and release v4 earlier than I anticipated.
> 
> A few minor changes are required, mostly around trailing commas. This would make it easier to merge and auto-merge PRs that add new imports later.

Absolutely, I'm moving so much between different languages that I didn't think about adding trailing commas to the imports. I will update!

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/requirements.txt
Comment (pergardebrink) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/requirements.txt:3]: I agree! And having all constructs available right away without needing to add new dependencies for each AWS service is so much more convenient.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/generate_pipeline_stacks.py
Comment (pergardebrink) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/generate_pipeline_stacks.py:37]: To be honest, I could have used the `DefaultStackSynthesizer` with the `generate_bootstrap_version_rule` parameter set to `False`, but since there are no assets in this stack, it makes more sense to use a synthesizer that doesn't require bootstrap in the first place.

**Comment by lpipiska** [comment]
Doesn't it make sense to update dependency of CodeBuild image to the latest version STANDARD_7_0 due to end of life for NodeJS?
STANDARD_6_0 uses NodeJS v16 which has end of life Sept 2023 (5 months from now).
STANDART_7_0 uses NodeJS v18 which has end of life in 2 years time.

**Comment by javydekoning** [comment]
> Doesn't it make sense to update dependency of CodeBuild image to the latest version STANDARD_7_0 due to end of life for NodeJS? STANDARD_6_0 uses NodeJS v16 which has end of life Sept 2023 (5 months from now). STANDART_7_0 uses NodeJS v18 which has end of life in 2 years time.

As discussed above, we would like to move the default CodeBuild image to `adfconfig.yml`. This would make it configurable and makes upgrading ADF easier going forward.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py
Comment (sbkok) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs/adf_codebuild.py:25]: Yes, that would be helpful. Could you change this back to the original default, then we can move that logic elsewhere later.

```suggestion
DEFAULT_CODEBUILD_IMAGE = "UBUNTU_14_04_PYTHON_3_7_1"
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/requirements.txt
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/requirements.txt:2]: Apologies for the delayed second review, in the meanwhile a few new CDK releases were published, could you update this to:

```suggestion
aws-cdk-lib==2.75.1
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml:682]:
```suggestion
                - npm install cdk@2.75 -g -y --quiet --no-progress
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:666]:
```suggestion
                - npm install cdk@2.75 -g -y --quiet --no-progress
```

**Review by sbkok — CHANGES_REQUESTED** [review]
Only a handful of minor changes to be made before we can merge this.
If you don't have bandwidth to apply these, please let me know and I can pick it up from here.

**Review by sbkok — APPROVED** [review]
LGTM, thank you for contributing this!

**Review by sbkok — DISMISSED** [review]

**Comment by javydekoning** [comment]
Looks like this impacts our CDK Constructs (see Tox failure). Also, not all of our CDK Constructs have tests.

https://github.com/awslabs/aws-deployment-framework/tree/master/src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/cdk/cdk_constructs

All though it looks like there is not much work needed: https://github.com/aws/aws-cdk-rfcs/blob/master/text/0192-remove-constructs-compat.md#release-notes

@sbkok thoughts?

**Comment by pergardebrink** [comment]
I see that the CDK was bumped since I created the pull request initially. If we bump it again, above 2.76, we would get STANDARD_7_0 support for ADF pipelines available, which would be really nice: https://github.com/aws/aws-cdk/releases/tag/v2.76.0

If no-one has the possibility to take a look at the failing test, I might be able to in the coming days but cannot promise anything.

**Comment by sbkok** [comment]
I am looking into how we can fix this atm.
The source of the problem is the default image that we relied on before.
The previous image, `UBUNTU_14_04_PYTHON_3_7_1` is not available when we upgrade to CDK v2.

So either way, we will need to introduce a breaking change to move forward.
I will push the changes into this PR once I found a solution that works.

**Review by sbkok — DISMISSED** [review]

**Comment by sbkok** [comment]
@javydekoning do you know if we can disable the cfnlint step for the time being? It seems it uses an old version in Megalinter (0.76.x) while 0.77.2 adds support for Python 3.10 which it complains about. Or might there be another way to fix this?

Alternatively, we could also disable cfnlint in Megalinter, as that is also executed as part of tox (which I was able to update to the version that works in this PR). Would you have time to look into this?

**Review by javydekoning — APPROVED** [review]
Approved, I can temporarely disable cfn-lint inMegalinter in a seperate PR.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml
Comment (pergardebrink) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml:682]: Is this package even needed here? 

I was trying to re-integrate the contents of this pull request into our testing ADF environment, but it failed due to the package cdk not found in version 2.79. For some reason, it seems that there's no 2.79 version published (as of 2023-05-15, see here: https://www.npmjs.com/package/cdk?activeTab=versions) but there's a 2.79 version of aws-cdk and the other cdk libraries: https://www.npmjs.com/package/aws-cdk-lib?activeTab=versions. 

Could we perhaps remove it if we don't need it (or if we need it, we might need to pick another package, should it be aws-cdk)?

Comment (javydekoning) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml:682]: I think you are correct, we need to use `aws-cdk`. The `cdk` package only links to `aws-cdk/bin/cdk` anyway.

See: https://github.com/aws/aws-cdk/blob/main/packages/cdk/bin/cdk

Comment (pergardebrink) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml:682]: I tried to understand what we actually needed the npm cdk package for here in the first place though? 

We are running some python scripts in this codebuild projects buildspec definition, but never calling the cdk cli or any JavaScript code as far as I can see, so I'm not sure I understand the need for the npm cdk package (or aws-cdk) either way in the first place? Or are we calling some cdk cli indirectly from the python scripts?

Comment (javydekoning) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml:682]: That is probably a left over from #424 and we no longer need `cdk` here indeed. @StewartW is that correct?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml
Comment (javydekoning) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:666]: Starting review, asking for change.

**Review by javydekoning — CHANGES_REQUESTED** [review]

**Comment by javydekoning** [comment]
> @javydekoning do you know if we can disable the cfnlint step for the time being? It seems it uses an old version in Megalinter (0.76.x) while 0.77.2 adds support for Python 3.10 which it complains about. Or might there be another way to fix this?

This is complete and merged to main.

**Comment by pergardebrink** [comment]
> @pergardebrink if you sync your branch from our main branch, remove the `cdk` install from `global.yml`, and change it to `aws-cdk` in `pipeline_management.yml` I think we are good to go on this.
> 
> Thanks a lot!

You mean merge the master branch into this/my branch? Sure! And I'll fix cdk => aws-cdk!

**Review by javydekoning — APPROVED** [review]

**Comment by javydekoning** [comment]
Thanks @pergardebrink , LGTM now!

@StewartW can you review as well?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml
Comment (StewartW) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/global.yml]: That's correct. 👍

**Comment by alFReD-NSH** [comment]
@StewartW I'm hoping this is not forgotten.

Currently CDK v1 only supports STANDARD_5_0 codebuild which only support nodejs 14.

Nodejs 14 now reached EOD and CDK v1 will reach EOD next week. It would really nice if we could speed this PR. cc @javydekoning

**Review by StewartW — APPROVED** [review]

**Comment by StewartW** [comment]
Looks good.

**Review by sbkok — APPROVED** [review]
Thank you, LGTM!

**Comment by markjl-netlogix** [comment]
Hi

This was merged in June. When can we expect a release that contains this merge?

**Comment by arifakkermans** [comment]
> Hi
> 
> This was merged in June. When can we expect a release that contains this merge?

I'm also waiting for a new release that includes the CDKv2 changes  @sbkok @javydekoning 

Is there an ETA on this?

**Comment by markjl-netlogix** [comment]
any progress on a release that contains CDKv2 changes?

**Comment by sbkok** [comment]
Apologies for the delayed response, new messages on closed PRs/Issues are hard to spot.
As you might have noticed, we have opened a number of PRs for the v4.0.0 release.
As the CDK v2 contains a breaking change, it will be part of the next major version release of ADF (v4).

There are a few more issues that need to be fixed before we can release.
We hope to get this done in the next couple of weeks.

**Comment by markjl-netlogix** [comment]
Hi @sbko

Any update on v4 release?

**Comment by sbkok** [comment]
Hi @markjl-netlogix,

I did not see your message until now, apologies. Unfortunately we ran into a number of issues that were hard to debug. Hence we did not release the new version just yet.

I appreciate your patience and understanding as our development team is working diligently to ensure the highest quality for this release.

Currently, we are in the final testing phase, and if no major issues arise, we are confident that the v4.0 release will be available within the next few weeks. Rest assured that our team is committed to delivering this update as soon as possible while maintaining the highest standards of quality and reliability.

We understand the importance of this release for our users, and we are eager to share the new features and improvements with you. Please stay tuned for further updates, and thank you once again for your continued support.

Best regards, Simon

**Pull request merged**
